# 4.8 Grouping Data & Aggregating Variables

## Contents List:

- Import libraries and orders_products_merged_updated.pkl

- Find aggregate mean using groupby

- Create loyalty_flag using transform() and loc()

- Check basic statistics of price using the loyalty_flag

- Create spending_flag

- Create frequency_flag
    - Check missing values

- Export orders_products_merged_updated_2.pkl

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import orders_products_merged_updated.pkl as dataframe

path = r'C:\Users\susan\OneDrive\Desktop\Data Analytics Program\June 2023 Instacart Basket Analysis'

In [3]:
df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [4]:
# Check shape

df_orders_products_merged.shape

(32404859, 18)

In [5]:
# Check output

df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid range product   
1         Soda        77              7     9.0   both  Mid range product   
2         Soda        77              7     9.0   both  Mid range product   
3         Soda        77              7     9.0   both  Mid range product   
4         Soda        77              7     9.0   both  Mid range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

# Task 4.8 Step 2: Find the aggregated mean of the 'order_number' column grouped by 'department_id'

In [6]:
df_orders_products_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3: Compare results for the entire dataframe with the results of the subset 

## The results for the entire dataframe include all 21 departments, whereas the results for the subset included only 8 departments. Also, the mean amounts are slightly different for the common departments between the entire dataframe and the subset because more data was used to calculate the mean.

# Step 4: Create loyalty flag using the transform() and loc() functions

In [7]:
# Create 'max_order' column

df_orders_products_merged['max_order'] = df_orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [10]:
# Check output

df_orders_products_merged.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid range product   
1          Soda        77              7     9.0   both  Mid range product   
2          Soda        77              7     9.0   both  Mid range product   
3          Soda        77              7     9.0   both  Mid range product   
4          Soda        77              7     9.0   both  Mid range product   
5          Soda        77              7     9.0   both  Mid range product   
6          Soda        77              7     9.0   both  Mid range product   
7          Soda        77              7     9.0   both  Mid range product   
8          Soda        77              7     9.0   both  Mid range product   
9          Soda        77              7     9.0   both  Mid range produ

In [9]:
# Check shape

df_orders_products_merged.shape

(32404859, 19)

In [11]:
# Create loyalty flag using loc()

df_orders_products_merged.loc[df_orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
df_orders_products_merged.loc[(df_orders_products_merged['max_order'] <= 40) & (df_orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
df_orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [15]:
df_orders_products_merged[['user_id', 'max_order', 'loyalty_flag']].head(30)

user_id  max_order      loyalty_flag
0         1         10      New customer
1         1         10      New customer
2         1         10      New customer
3         1         10      New customer
4         1         10      New customer
5         1         10      New customer
6         1         10      New customer
7         1         10      New customer
8         1         10      New customer
9         1         10      New customer
10       15         22  Regular customer
11       15         22  Regular customer
12       15         22  Regular customer
13       15         22  Regular customer
14       15         22  Regular customer
15       19          9      New customer
16       19          9      New customer
17       19          9      New customer
18       21         33  Regular customer
19       31         20  Regular customer
20       31         20  Regular customer
21       43         11  Regular customer
22       43         11  Regular customer
23       52         27  Regular customer
24       52         27  Regular customer
25       52         27  Regular customer
26       52         27  Regular customer
27       52         27  Regular customer
28       52         27  Regular customer
29       52         27  Regular customer

# Step 5: Use loyalty flag to check basic statistics of the product prices for each loyalty category

In [16]:
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## The mean product price for loyal customers is lower than the price for regular and new customers. There is likely something incorrect about a max price of 99999 for all 3 categories, but we would need another source to verify product prices in our data

# Step 6: Create a spending flag for each user

In [17]:
# Create 'mean_price' column

df_orders_products_merged['mean_price'] = df_orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [18]:
# Check output

df_orders_products_merged.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid range product   
1          Soda  ...              7     9.0    both  Mid range product   
2          Soda  ...              7     9.0    both  Mid range product   
3          Soda  ...              7     9.0    both  Mid range product   
4          Soda  ...              7     9.0    both  Mid range product   
5          Soda  ...              7     9.0    both  Mid range product   
6          Soda  ...              7     9.0    both  Mid range product   
7          Soda  ...              7     9.0    both  Mid range product   
8          Soda  ...              7     9.0    both  Mid range product   
9          Soda  ...              7     9.0    both  Mid range product   
10         Soda  ...              7   

In [19]:
# Check shape

df_orders_products_merged.shape

(32404859, 21)

In [20]:
# Create spending flag using loc()

df_orders_products_merged.loc[df_orders_products_merged['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [21]:
df_orders_products_merged.loc[df_orders_products_merged['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [22]:
df_orders_products_merged['spending_flag'].value_counts(dropna = False)

Low spender     31770742
High spender      634117
Name: spending_flag, dtype: int64

In [24]:
# Check output

df_orders_products_merged.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0          Soda  ...     9.0    both  Mid range product  Regularly busy   
1          Soda  ...     9.0    both  Mid range product  Regularly busy   
2          Soda  ...     9.0    both  Mid range product  Regularly busy   
3          Soda  ...     9.0    both  Mid range product      Least busy   
4          Soda  ...     9.0    both  Mid range product      Least busy   
5          Soda  ...     9.0    both  Mid range product  Regularly busy   
6          Soda  ...     9.0    both  Mid range product  Regularly busy   
7          Soda  ...     9.0    both  Mid range product  Regularly busy   
8          Soda  ...     9.0    both  Mid range product  Regularly busy   
9          Soda  ...     9.0    both  Mid range product      Least busy   
10         Soda  ...     9.

# Step 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the 'days_since_prior_order' column

In [25]:
# Create median_days_since_prior_order column

df_orders_products_merged['median_days_since_prior_order'] = df_orders_products_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
# Check output

df_orders_products_merged.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  _merge    price_range_loc     busiest_day  \
0          Soda  ...    both  Mid range product  Regularly busy   
1          Soda  ...    both  Mid range product  Regularly busy   
2          Soda  ...    both  Mid range product  Regularly busy   
3          Soda  ...    both  Mid range product      Least busy   
4          Soda  ...    both  Mid range product      Least busy   
5          Soda  ...    both  Mid range product  Regularly busy   
6          Soda  ...    both  Mid range product  Regularly busy   
7          Soda  ...    both  Mid range product  Regularly busy   
8          Soda  ...    both  Mid range product  Regularly busy   
9          Soda  ...    both  Mid range product      Least busy   
10         Soda  ...    both  Mid range product  Regularly busy   
11         Soda  ...    both  Mid range product 

In [27]:
# Check shape

df_orders_products_merged.shape

(32404859, 23)

In [28]:
# Create frequency flag using loc()

df_orders_products_merged.loc[df_orders_products_merged['median_days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [30]:
df_orders_products_merged.loc[(df_orders_products_merged['median_days_since_prior_order'] > 10) & (df_orders_products_merged['median_days_since_prior_order'] <= 20), 'frequency_flag'] = 'Regular customer'

In [33]:
df_orders_products_merged.loc[df_orders_products_merged['median_days_since_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [34]:
df_orders_products_merged['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [35]:
# Check for missing values

df_orders_products_merged.isnull().sum()

order_id                               0
user_id                                0
order_number                           0
orders_day_of_week                     0
order_hour_of_day                      0
days_since_prior_order           2076096
product_id                             0
add_to_cart_order                      0
reordered                              0
product_name                           0
aisle_id                               0
department_id                          0
prices                                 0
_merge                                 0
price_range_loc                        0
busiest_day                            0
busiest_days                           0
busiest_period_of_day                  0
max_order                              0
loyalty_flag                           0
mean_price                             0
spending_flag                          0
median_days_since_prior_order          5
frequency_flag                         5
dtype: int64

In [36]:
# View the 5 missing values

df_orders_products_merged[df_orders_products_merged['frequency_flag'].isnull() == True]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  busiest_day  busiest_days busiest_period_of_day  \
13645692  Mid range product  Busiest day  Busiest days        Average orders   
17251990  Mid range product  Busiest day  Busiest days        Average orders   
17622767  Low range product  Busiest day  Busiest days        Average orders   
24138593  Mid range product  Busiest day  Busiest days        Average orders   
25880002  Low range product  Busiest day  Busiest days        Average orders   

         max_order  loyalty_flag mean_price spending_flag  \
13645692         1  New customer       7.42   Low spender   
17251990         1  New customer       7.42   Low spender   
17622767         1  New customer       7.42   Low spender   
24138593         1  New customer       7.42   Low spender   
25880002         1  New customer       7.42   Low spender   

          median_days_since_prior_order frequency_flag  
13645692                            NaN            NaN  
17251990                            NaN            NaN  
17622767                            NaN            NaN  
24138593                            NaN            NaN  
25880002                            NaN            NaN  

[5 rows x 24 columns]

## The 5 missing values in the frequency_flag are from a new customer who placed only 1 order and never re-ordered again. So their days_since_prior_order column was empty. This customer can be categorized as a 'Non-frequent customer'

In [37]:
# Change the missing values to 'Non-frequent customer'

df_orders_products_merged['frequency_flag'].fillna('Non-frequent customer', inplace = True)

In [38]:
# Check frequency_flag value counts

df_orders_products_merged['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: frequency_flag, dtype: int64

In [39]:
# Check shape of dataframe before export

df_orders_products_merged.shape

(32404859, 24)

In [40]:
# Check output of dataframe before export

df_orders_products_merged.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   
20                     8.0         196                 14          1   
2

# Export df_orders_products_merged

In [41]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))